base.py

```diff
@@ -178,45 +192,144 @@ def run(self):
             exit_code = 1
             print('\nThe internet connection was lost')
 
+        except APIError as e:
+            exit_code = 1
+            print('\nAPIError: {}'.format(e.msg))
+
         finally:
-            for post_test_case in self._post_test_cases:
-                method = getattr(self, post_test_case)
-                print('Running clean up {}...'.format(post_test_case))
-                method()
+            self.handle_env(self.tear_down)
+            self.handle_connections(self._close_connections)
 
-            if os.path.isfile('/.docker'):
-                self.xvfb.stop()
+            return exit_code
 
-        return exit_code
+    @staticmethod
+    def handle_env(func):
+        print('Starting {}...'.format(func.__name__), end=' ', flush=True)
+        try:
+            func()
+
+        except APIError as e:
+            print(' failed')
+            print('APIError: {}'.format(e.msg))
+
+            return False
+
+        print(' done')
+
+        return True
+
+    def handle_connections(self, connections_lst):
+        for conn in connections_lst:
+            method = getattr(self, conn)
+            output = ' '.join(conn.capitalize().split('_'))
+            print(output)
+            connected = method()
+
+            if not connected:
+                return False
+
+        return True
+
+    def connect_to_browser(self):
+        if os.path.isfile('/.docker'):
+            # xvfb wrapper starting
+            print('Using Xvfb')
+            self.xvfb = Xvfb()
+
+            try:
+                self.xvfb.start()
+
+            except Exception:
+                return False
+
+        options = Options()
+        options.add_argument('--no-sandbox')
+
+        try:
+            self.browser = Browser('chrome', headless=False, options=options,
+                                   wait_time=30,
+                                   executable_path='./drivers/chromedriver')
+
+        except Exception:
+            print('\nFailed to connect with browser.')
+            return False
+
+        self.browser.driver.implicitly_wait(self.sticky_timeout)
+        self.browser.driver.set_page_load_timeout(self.page_load_timeout)
+        self.browser.driver.set_window_size(*self.browser_window_size)
+        self.browser.visit(self.addr)
+
+        return True
+
+    def close_browser_connection(self):
+        if os.path.isfile('/.docker'):
+            self.xvfb.stop()
+
+        return True
+
+    def setup(self):
+        pass
+
+    def tear_down(self):
+        pass
 
 
 class RocketChatTestCase(SplinterTestCase):
-    def __init__(self, addr, username, password, create_test_user=True,
-                 **kwargs):
+    def __init__(self, addr, username, password, expected_rooms=None, **kwargs):
```

Comment:
String includes 80 symbols. It's longer than the PEP8 standard. Yes it's only 1 symbol, but you can cut it and be standard compliance